bin/brew

```diff
@@ -1,42 +1,98 @@
+  }
+
+  rl_canonicalize()
+  {
+    [ "$(basename "$*")"- = "."- ] || [ -"$(basename "$*")"- = -".."- ] &&
+      rl_canon=$(cd "$(pwd -P)/$(basename "$*")"; pwd -P) ||
+      rl_canon=$(echo $(pwd -P)/$(basename "$*") | sed 's|//|/|g')
+    rl_canonical=$(echo "$rl_canon" | sed 's|//|/|g')
+    echo "$rl_canonical"
+  }
+
+  rl_no_dir()
+  {
+    echo "$*" && return 0
+  }
+
+  rl_not_link()
+  {
+    rl_canonicalize "$*" && exit 0
+  }
+
+  rl_try(){
+    rl_cur_dir=$(dirname "$*")
+    rl_cur_base=$(basename "$*")
+
+    cd "$rl_cur_dir" 2>/dev/null || rl_no_dir "$*"
+    [ -e "$rl_cur_base" ] || rl_canonicalize "$(pwd -P)/$*"
+    [ -L "$rl_cur_base" ] || rl_not_link "$*"
+
+    rl_rl=$(rl_rl "$rl_cur_base")
+    rl_try "$rl_rl"
+  }
+
+  rl_try "$*"
 }
 
-BREW_FILE_DIRECTORY="$(chdir "${0%/*}" && pwd -P)"
-HOMEBREW_BREW_FILE="$BREW_FILE_DIRECTORY/${0##*/}"
+HOMEBREW_BREW_FILE=$(realpath "$0")
+BREW_FILE_DIRECTORY=$(dirname "$HOMEBREW_BREW_FILE")
+BREW_LIBRARY_DIRECTORY="$(cd "$BREW_FILE_DIRECTORY/../Library"; pwd -P)"
+HOMEBREW_PREFIX=$(dirname "$BREW_FILE_DIRECTORY")
+
+export BREW_FILE_DIRECTORY
+export BREW_LIBRARY_DIRECTORY
+export HOMEBREW_PREFIX
+export HOMEBREW_BREW_FILE
+
+case "$1" in
+
+          update) shift
+              exec "$BREW_LIBRARY_DIRECTORY/Homebrew/cmd/update"     "$@" ;;
+
+      shellutils) shift
+              exec "$BREW_LIBRARY_DIRECTORY/Homebrew/cmd/shellutils" "$@" ;;
+
+        --prefix) shift
+              [ $# -eq 0 ] && echo $HOMEBREW_PREFIX && exit 0
 
-if [ -L "$HOMEBREW_BREW_FILE" ]
-then
-  BREW_SYMLINK="$(readlink "$HOMEBREW_BREW_FILE")"
-  BREW_SYMLINK_DIRECTORY="$(dirname "$BREW_SYMLINK")"
-  BREW_FILE_DIRECTORY="$(chdir "$BREW_FILE_DIRECTORY" &&
-                         chdir "$BREW_SYMLINK_DIRECTORY" && pwd -P)"
-fi
+              for each in "$@"; do
+                incellar=$(ls -1d $HOMEBREW_PREFIX/Cellar/$each/* 2>/dev/null)
+                [ -z "$incellar" ] && echo "$HOMEBREW_PREFIX/opt/$each" ||
+                  echo "$cellar"
+                shift
+              done
 
-BREW_LIBRARY_DIRECTORY="$(chdir "$BREW_FILE_DIRECTORY"/../Library && pwd -P)"
+              exit 0 ;;
 
-# Users may have these set, pointing the system Ruby
-# at non-system gem paths
+           *) tapcommand=$(ls $HOMEBREW_PREFIX/*/*/*/*/*/brew-$1 2>/dev/null)
+              [ ! -z $tapcommand ] && shift && exec "$tapcommand" "$@"
+
+              externalcommand=$(which brew-$1) && shift &&
+                exec "$externalcommand" "$@" || continue ;;
+
+        ''|*) continue ;;
+esac
+
+# Users may have these set, pointing the system Ruby at non-system gem paths
 unset GEM_HOME
 unset GEM_PATH
 
-if [ -z "$HOMEBREW_DEVELOPER" ]
-then
-  unset HOMEBREW_RUBY_PATH
-fi
-
-if [ -z "$HOMEBREW_RUBY_PATH" ]
-then
-  if [ "$(uname -s)" = "Darwin" ]
-  then
-    HOMEBREW_RUBY_PATH="/System/Library/Frameworks/Ruby.framework/Versions/Current/usr/bin/ruby"
-  else
-    HOMEBREW_RUBY_PATH="$(which ruby)"
-  fi
-fi
```

Comment:
Similarly let's remove all the changes to this file for now except for a special-case for `update`.